### PR TITLE
VectorizedMarkovMessenger proposal

### PR DESCRIPTION
**Pull request by ordabayevy**
`VectorizedMarkovMessenger` proposal to address #2687 and as a Pyro interface for pyro-ppl/funsor#398. The goal was to have the model syntax for `pyro.vectorized_markov` as close as possible to `pyro.markov`.

`pyro.vectorized_markov` iterates over two `for` loops for `history > 0`:

1. `for i in range(self.history)` - initializes first `history` steps where `i` is int (sites `x_0` and `y_0` below)
2. `for i in range(self.history+1)` builds necessary factors for the `modified_partial_sum_product` (sites `x_tensor([1, 2])` and `y_tensor([1, 2])`) and creates `plate_to_step` (`step = {"x_tensor([0, 1])":"x_tensor([1, 2])"}`) dict while returning vectorized `i` indices

```diff
data = torch.ones(3, dtype=torch.float)

def model(data):

    init = pyro.param("init", lambda: torch.rand(3), constraint=constraints.simplex)
    trans = pyro.param("trans", lambda: torch.rand((3, 3)), constraint=constraints.simplex)
    locs_mean = pyro.param("locs_mean", lambda: torch.ones((3,)))
    locs = pyro.sample("locs", dist.Normal(locs_mean, 1.).to_event(1))

-   for i in pyro.markov(range(len(data))):
+   for i in pyro.vectorized_markov(name="time", size=len(data), dim=-1):
        x_curr = pyro.sample("x_{}".format(i), dist.Categorical(
            init if isinstance(i, int) and i < 1 else trans[x_prev]),

        pyro.sample("y_{}".format(i), dist.Normal(Vindex(locs)[..., x_curr], 1.), obs=data[i])
        x_prev = x_curr
```

<details><summary>pyro.markov trace</summary>

```
Trace Shapes:                
 Param Sites:                
         init             3  
        trans           3 3  
    locs_mean             3  
Sample Sites:                
    locs dist             | 3
        value             | 3
     log_prob             |  
     x_0 dist             |  
        value   3 1 1 1 1 |  
     log_prob   3 1 1 1 1 |  
     y_0 dist   3 1 1 1 1 |  
        value             |  
     log_prob   3 1 1 1 1 |  
     x_1 dist   3 1 1 1 1 |  
        value 3 1 1 1 1 1 |  
     log_prob 3 3 1 1 1 1 |  
     y_1 dist 3 1 1 1 1 1 |  
        value             |  
     log_prob 3 1 1 1 1 1 |  
     x_2 dist 3 1 1 1 1 1 |  
        value   3 1 1 1 1 |  
     log_prob 3 3 1 1 1 1 |  
     y_2 dist   3 1 1 1 1 |  
        value             |  
     log_prob   3 1 1 1 1 |  
```

</details>

<details><summary>vectorized_trace</summary>

```
        Trace Shapes:                  
         Param Sites:                  
                 init               3  
                trans             3 3  
            locs_mean               3  
        Sample Sites:                  
            locs dist               | 3
                value               | 3
             log_prob               |  
             x_0 dist               |  
                value     3 1 1 1 1 |  
             log_prob     3 1 1 1 1 |  
             y_0 dist     3 1 1 1 1 |  
                value               |  
             log_prob     3 1 1 1 1 |  
x_tensor([1, 2]) dist   3 1 1 1 1 2 |  
                value 3 1 1 1 1 1 1 |  
             log_prob 3 3 1 1 1 1 2 |  
y_tensor([1, 2]) dist 3 1 1 1 1 1 2 |  
                value             2 |  
             log_prob 3 1 1 1 1 1 2 |  
```

</details>

<details><summary>vectorized_trace.nodes["x_tensor([1, 2])"]["funsor"]["log_prob"].inputs</summary>

```
OrderedDict([('x_tensor([1, 2])', Bint[3, ]),
             ('x_tensor([0, 1])', Bint[3, ]),
             ('time', Bint[2, ])])
```

</details>

- `history = 0` behaves like a `pyro.plate` with an empty step in `plate_to_step` and no sub-sampling

```diff
def model(data):

    p = pyro.param("probs", lambda: torch.rand(3), constraint=constraints.simplex)
    locs_mean = pyro.param("locs_mean", lambda: torch.ones((3,)))
    locs = pyro.sample("locs", dist.Normal(locs_mean, 1.).to_event(1))

-   with pyro.plate("time", len(data), dim=-1) as i:
+   for i in pyro.vectorized_markov(name="time", size=len(data), dim=-1, history=0):
        x = pyro.sample("x", dist.Categorical(p))
        pyro.sample("y", dist.Normal(Vindex(locs)[..., x], 1.), obs=data[i])
```

**Tasks**

- [x] Create `plate_to_step` dict (e.g. `plate_to_step["time"] = {"x_prev": "x_curr"} (pair coded with @eb8680 )
- [x] Return vectorized time indices in the for loop
- [x] Handle initialization (time zero) syntax
- [x] Add tests

**Tests** - compare `factor.inputs` and `to_data(factor)` between factors created by `pyro.markov` model vs hand-unrolled factors created by `pyro.vectorized_markov` model
- [x] `model_0` outer plate `sequences`
- [x] `model_[1-4]` somewhat similar to HMM examples
- [x] `model_5` history = 2
- [x] `model_6` - `x_trans` is time dependent
- [x] `model_7` - correct `step` for `x_prev->w_curr` and `w_prev->x_curr` dependency
- [x] `model_8` - two independent time dims `weeks` and `days`

### CODE REVIEW

**Comment by ordabayevy** [comment]
@eb8680 I have created some tests that compare `factor.inputs` and `to_data(factor)` between factors created by `pyro.markov` and hand-unrolled factors created by `pyro.vectorized_markov`. `plate_to_step` dict also looks correct but those are not included in the test. Do you have any suggestions for other tests?

Current version of `VectorizedMarkovMessenger` should work for all of my models. But I'm happy to tweak it more if there are other use cases that require adjustments.

**Review by eb8680 — COMMENTED** [review]
This is looking pretty good!  I have some comments, mostly nits and requests for documentation.  I'd also like to think more about the model interface and see if there's a more principled solution for matching variables across timesteps than the string suffix heuristic we're using, but that doesn't necessarily have to block this PR.

Another test case that might be useful for your problem is a model with multiple independent time axes.

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (ordabayevy) [reply pyro/contrib/funsor/handlers/named_messenger.py:223]: - The initial loop `for i in range(self.history)` creates `x_0` (shape 3,1,1,1,1) variable which initiates the markov chain (same in `pyro.markov`)
- The rest of the chain is vectorized. `x_tensor([1, 2])` factor (shape 3,3,1,1,1,1,2) corresponds to two factors `x_1` (3,3,1,1,1,1) and `x_2` (3,3,1,1,1,1) in `pyro.markov`
- The benefit of dealing with the initial state is that the following vectorized loop `for i in range(self.history+1)` returns indices of the same length: tensor([0, 1]) (corresponding to time prev `self.markov_iter=0`) and tensor([1, 2]) (corresponding to time curr `self.markov_iter=1`)

This also allows to do one to one mapping between `pyro.markov` factors and hand-unrolled `pyro.vectorized_markov` factors as I do in the test. It is easier to see the correspondence by directly comparing two traces:

<details><summary>pyro.markov trace</summary>

```
Trace Shapes:                
 Param Sites:                
         init             3  
        trans           3 3  
    locs_mean             3  
Sample Sites:                
    locs dist             | 3
        value             | 3
     log_prob             |  
     x_0 dist             |  
        value   3 1 1 1 1 |  
     log_prob   3 1 1 1 1 |  
     y_0 dist   3 1 1 1 1 |  
        value             |  
     log_prob   3 1 1 1 1 |  
     x_1 dist   3 1 1 1 1 |  
        value 3 1 1 1 1 1 |  
     log_prob 3 3 1 1 1 1 |  
     y_1 dist 3 1 1 1 1 1 |  
        value             |  
     log_prob 3 1 1 1 1 1 |  
     x_2 dist 3 1 1 1 1 1 |  
        value   3 1 1 1 1 |  
     log_prob 3 3 1 1 1 1 |  
     y_2 dist   3 1 1 1 1 |  
        value             |  
     log_prob   3 1 1 1 1 |  
```

</details>

<details><summary>vectorized_trace</summary>

```
        Trace Shapes:                  
         Param Sites:                  
                 init               3  
                trans             3 3  
            locs_mean               3  
        Sample Sites:                  
            locs dist               | 3
                value               | 3
             log_prob               |  
             x_0 dist               |  
                value     3 1 1 1 1 |  
             log_prob     3 1 1 1 1 |  
             y_0 dist     3 1 1 1 1 |  
                value               |  
             log_prob     3 1 1 1 1 |  
x_tensor([1, 2]) dist   3 1 1 1 1 2 |  
                value 3 1 1 1 1 1 1 |  
             log_prob 3 3 1 1 1 1 2 |  
y_tensor([1, 2]) dist 3 1 1 1 1 1 2 |  
                value             2 |  
             log_prob 3 1 1 1 1 1 2 |  
```

</details>

Comment (ordabayevy) [reply pyro/contrib/funsor/handlers/named_messenger.py:223]: -  The first loop is for `p(x_0)` 
- The second vectorized loop is for `PROD[t=1...T] p(x_t|x_t-1)` or `p(x_1|x_0)p(x_2|x_1)...`

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (ordabayevy) [reply pyro/contrib/funsor/handlers/named_messenger.py:286]: Okay, it does now. (It seems to work both ways, I'll have to think what's the difference)

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (ordabayevy) [reply tests/contrib/funsor/test_vectorized_markov.py:127]: See `model_6`. I'll try to come up with another test that also uses time indices.

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (ordabayevy) [reply tests/contrib/funsor/test_vectorized_markov.py:367]: It checks the `step` information now. Note: instead of the `plate_to_step` there is a `time` node in the trace with `step` information in the `infer` dict.

**Review by ordabayevy — COMMENTED** [review]
- Removed any changes from `trace_struct.py`, `traceenum_elbo.py` and `poutine/trace_messenger.py` files
- Reworked `_markov_step` and how `step` is determined which supports history > 1 tested by `model_5`
- Added a test for `use_replay`

There are significant changes to `step` which we should discuss. It is a frozenset of tuples. Each tuple is sequence of names `("x_0", "x_tensor([0,1])", "x_tensor([1,2])")` corresponding to (init,prev,curr). I think we should add `init` to `modified_partial_sum_product` too and combine it with reduced `trans` like [here](https://github.com/pyro-ppl/pyro/blob/68aad8c8e94dadecf522e3836550b6f5812be852/pyro/distributions/hmm.py#L343).

Yet to add docstrings.

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (fehiepsi) [inline tests/contrib/funsor/test_vectorized_markov.py:150]: Awesome! IIUC, this api supports history>=2. Is there any requirement/restriction for the code to achieve that?

@eb8680 Would this be a solution for https://github.com/pyro-ppl/numpyro/issues/702? Which API should I use for the future: `scan` or `vectorized_markov`? :D

Comment (ordabayevy) [reply tests/contrib/funsor/test_vectorized_markov.py:150]: Yup, it does! Here is an example written using both `pyro.markov` and `pyro.vectorized_markov`:

```py
def model_5(data, history=2, vectorized=True):
    x_dim, y_dim = 3, 2
    pyro.set_rng_seed(0)
    pyro.get_param_store().clear()
    x_init = pyro.param("x_init", lambda: torch.rand(x_dim), constraint=constraints.simplex)
    x_init_2 = pyro.param("x_init_2", lambda: torch.rand(x_dim, x_dim), constraint=constraints.simplex)
    x_trans = pyro.param("x_trans", lambda: torch.rand((x_dim, x_dim, x_dim)), constraint=constraints.simplex)
    y_probs = pyro.param("y_probs", lambda: torch.rand(x_dim, y_dim), constraint=constraints.simplex)

    x_prev = x_prev_2 = None
    markov_loop = \
        pyro.vectorized_markov(name="time", size=len(data), dim=-2, history=history) if vectorized \
        else pyro.markov(range(len(data)), history=history)
    for i in markov_loop:
        if isinstance(i, int) and i == 0:
            x_probs = x_init
        elif isinstance(i, int) and i == 1:
            x_probs = Vindex(x_init_2)[x_prev]
        else:
            x_probs = Vindex(x_trans)[x_prev_2, x_prev]

        x_curr = pyro.sample(
            "x_{}".format(i), dist.Categorical(x_probs),
            infer={"enumerate": "parallel"})
        with pyro.plate("tones", data.shape[-1], dim=-1):
            pyro.sample("y_{}".format(i), dist.Categorical(
                Vindex(y_probs)[x_curr]),
                obs=data[i])
        x_prev_2, x_prev = x_prev, x_curr
```

Comment (ordabayevy) [reply tests/contrib/funsor/test_vectorized_markov.py:150]: Sorry, I completely missed the reference to the code!

Comment (ordabayevy) [reply tests/contrib/funsor/test_vectorized_markov.py:150]: My hope is that any code that can be written with `pyro.markov` (using proper `init` probabilities) can be replaced with `pyro.vectorized_markov`.

Comment (eb8680) [reply tests/contrib/funsor/test_vectorized_markov.py:150]: > IIUC, this api supports history>=2. Is there any requirement/restriction for the code to achieve that?

Well, right now the corresponding tensor variable elimination algorithm for `history > 1` doesn't exist - the one in https://github.com/pyro-ppl/funsor/pull/398 only supports `history <= 1`.  The enumeration behavior in this PR should be general, though.

> Would this be a solution for pyro-ppl/numpyro#702? Which API should I use for the future: scan or vectorized_markov? :D

Once this code is finished, I think it might make sense to write an equivalent JAX-flavored version in terms of `scan` for NumPyro, which would be a realization of https://github.com/pyro-ppl/numpyro/issues/686.  That should be pretty easy after the details are worked out here. As @ordabayevy says, our design goal is to make `pyro.vectorized_markov` interchangeable with `pyro.markov` whenever that is mathematically permissible.

Comment (fehiepsi) [reply tests/contrib/funsor/test_vectorized_markov.py:150]: I see, thanks for explaining! This API should be convenient for many usage cases (though I guess `scan` is still useful when it is tricky to write vectorized code) and the code might look more intuitive with `for` loop. In addition, this does not require a scan loop to collect the values, which enables a performance boost.

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:25]: nit: remove these two lines from all of the models, they are handled in our pytest setup

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:451]: nit: change this to `vectorized_trace.nodes["weeks"]["value"]`

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:459]: nit: change this to `vectorized_trace.nodes["days"]["value"]`

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:328]: nit: change this to `vectorized_trace.nodes["time"]["value"]`

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/named_messenger.py:296]: nit: `correct` to `valid`

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/named_messenger.py:249]: nit: change this to `trace.nodes["time"]["value"]`

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/named_messenger.py:309]: Can you get rid of `keep` and all associated code for now? It's not clear that `keep=True` is compatible with vectorized enumeration. We can revisit this in future PRs if necessary.

File: pyro/contrib/funsor/handlers/named_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/named_messenger.py:316]: ditto: remove `keep`

**Review by eb8680 — COMMENTED** [review]
Still reviewing the tests, but here are some further comments on the latest changes.

Do you want to explain how you plan to use the new information in `step` here, or would it be easier to do over Zoom?

File: pyro/contrib/funsor/handlers/plate_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/plate_messenger.py:297]: Wow, this is awesome! Great that it turned out to be so simple!

Comment (ordabayevy) [reply pyro/contrib/funsor/handlers/plate_messenger.py]: There was one issue. Had to take out `range(self.history)` initialization steps out of `IndepMessenger`, otherwise `time` dim gets added to the `cond_indep_stack` and broadcasted. Also removed `MarkovMessenger` for now because there are two for loops.

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:271]: Can you add a couple new test cases to this list with the same models/variables/histories but with longer time axes (>50)? e.g.
```
(model_8, torch.ones((50, 4), dtype=torch.long), "wxy", 1)
```

File: tests/contrib/funsor/test_vectorized_markov.py
Comment (eb8680) [inline tests/contrib/funsor/test_vectorized_markov.py:373]: ditto: add a case with longer time axes:
```
(model_7, torch.ones(30), torch.zeros(50), "xy", "wz", 1)
```

File: pyro/contrib/funsor/handlers/plate_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/plate_messenger.py:324]: nit: check if `funsor_log_prob` has already been computed:
```
funsor_log_prob = msg["funsor"]["log_prob"] if "log_prob" in msg["funsor"] else \
    to_funsor(msg["fn"].log_prob(msg["value"]), output=funsor.Real)
```

File: pyro/contrib/funsor/handlers/plate_messenger.py
Comment (eb8680) [inline pyro/contrib/funsor/handlers/plate_messenger.py:178]: Have you run `make docs` locally to check that your docstrings render correctly?

**Review by eb8680 — COMMENTED** [review]
I only have a few more minor comments. The tests look good to me overall.

**Comment by ordabayevy** [comment]
Addressed all of the comments. Changed `model_0` to a model with an outer plate `sequences`, time dim `time`, and an inner plate `tones` because there wasn't any model with an outer plate.

For models with longer time added 50 and 100. Tried out 500 which has passed but it was slow so I removed it.

**Comment by ordabayevy** [comment]
Summary:
- added `_do_not_trace` logic to the `TraceMessenger`
- tried `re.sub` but didn't work (because of the spaces in the names? idk). Then borrowed an idea from a `removesuffix` function introduced in [Python 3.9](https://www.python.org/dev/peps/pep-0616/) 
- yielding `slice` object breaks the tests because indexing with `slice` and `torch.tensor` work differently especially in `Vindex`. Also yielding `torch.tensor` object still might be better because it is easier to manipulate it compared to `slice` object. As intermediary solution for loop yields `torch.tensor` and then in `_pyro_sample` swaps `tensor(..)` suffix with `slice(...)` suffix.

**Review by eb8680 — APPROVED** [review]
@ordabayevy sorry for the delay, I had some reviews to do over the weekend. I'm happy to merge this pending confirmation that the docs render correctly for you locally. Great job!

Per our discussion on Friday, the next steps are updating `funsor.sum_product.modified_partial_sum_product` to include `init` factors and integrating it into `pyro.contrib.funsor.infer.TraceEnum_ELBO` in place of `partial_sum_product`.

**Comment by ordabayevy** [comment]
@eb8680 thanks again for helping with this PR and reviewing it!

Here is the output of the `make docs`:

<details>

``` bash
make -C docs html
make[1]: Entering directory '/home/ordabayev/repos/pyro/docs'
Running Sphinx v3.3.0
/home/ordabayev/repos/pyro/docs/source/conf.py:209: RemovedInSphinx40Warning: The app.add_stylesheet() is deprecated. Please use app.add_css_file() instead.
  app.add_stylesheet('css/pyro.css')
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://pytorch.org/docs/master/objects.inv...
loading intersphinx inventory from http://funsor.pyro.ai/en/stable/objects.inv...
loading intersphinx inventory from https://optimized-einsum.readthedocs.io/en/stable/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/scipy/reference/objects.inv...
loading intersphinx inventory from https://biopython.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from https://horovod.readthedocs.io/en/stable/objects.inv...
intersphinx inventory has moved: http://funsor.pyro.ai/en/stable/objects.inv -> https://funsor.pyro.ai/en/stable/objects.inv
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 28 source files that are out of date
updating environment: [new config] 28 added, 0 changed, 0 removed
reading sources... [100%] primitives
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] primitives
generating indices... genindex py-modindex done
highlighting module code... [100%] torch.distributions.constraints
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/html.
make[1]: Leaving directory '/home/ordabayev/repos/pyro/docs'
```

</details>


> Per our discussion on Friday, the next steps are updating `funsor.sum_product.modified_partial_sum_product` to include init factors and integrating it into `pyro.contrib.funsor.infer.TraceEnum_ELBO` in place of `partial_sum_product.`

Sounds like a plan. First, I'll submit an updated `partial_unroll` to handle markov dimensions. That should help a lot with the following steps that you mentioned.

**Comment by eb8680** [comment]
> Here is the output of the make docs:

Can you open the generated HTML files on your computer and check that they look correct visually?

**Comment by ordabayevy** [comment]
@eb8680 had to fix few things like escaping `\` in the code block. Look ok to me now.

While looking at the HTML files I noticed that sphinx references in other `pyro.contrib.funsor` `Messenger`s are rendered incorrectly, e.g. ``:func:~`pyro.contrib.funsor.to_data` `` and ``:class:~`MarkovMessenger` `` should be fixed by ``:func:`~pyro.contrib.funsor.to_data` ``  and  ``:class:`~MarkovMessenger` `` in many places. But that's another PR I guess.

**Review by eb8680 — APPROVED** [review]

**Pull request merged**
